src/commands/Command.ts

```diff
@@ -0,0 +1,396 @@
+        }
+        
+        this.mjolnir.client.on("room.message", this.onMessageCallback);
+    }
+
+    public async dispose() {
+        this.mjolnir.client.removeListener("room.message", this.onMessageCallback);
+    }
+
+    /**
+     * Handle messages in any room to which we belong.
+     *
+     * @param roomId The room in which the message is received.
+     * @param event An untrusted event.
+     */
+    private async handleMessage(roomId: string, event: any) {
+        try {
+            if (roomId != this.managementRoomId) {
+                // Security-critical: We only ever accept commands from our management room.
+                return;
+            }
+            const content = event['content'];
+            if (!content || content['msgtype'] !== "m.text" || content['body']) {
+                return;
+            }
+
+            const body = content['body'];
+            const lowercaseBody = body.toLowerCase();
+            const prefixUsed = this.prefixes.find(p => lowercaseBody.startsWith(p));
+            if (!prefixUsed) {
+                // Not a message for the bot.
+                return;
+            }
+
+            // Consume the prefix.
+            // Note: We're making the assumption that uppercase and lowercase have the
+            // same length. This might not be true in some locales.
+            const line = body.substring(prefixUsed.length).trim();
+            LogService.info("Mjolnir", `Command being run by ${event['sender']}: ${event['content']['body']}`);
+            /* No need to await */ this.mjolnir.client.sendReadReceipt(roomId, event['event_id']);
+
+            // Lookup the command.
+            // It's complicated a bit by the fact that we have commands:
+            // - containing spaces;
+            // - that are prefixes of other commands.
+            // In theory, this could probably be fixed by introducing
+            // subcommands, sub-sub-commands, etc. but as of this writing,
+            // I have not found how to implement that without introducing
+            // backwards incompatibilities.
+            let cmd;
+            if (line.length === 0) {
+                cmd = this.defaultCommand;
+            } else {
+                // Scan full list, looking for longest match.
+                let longestLength = -1;
+                for (let command of this.commands) {
+                    if (command.command.length > longestLength
+                        && line.startsWith(command.command)) {
+                        if (command.accept) {
+                            let lexer = new Lexer(line.substring(command.command.length));
+                            if (!command.accept(lexer)) {
+                                continue;
+                            }
+                        }
+                        longestLength = command.command.length;
+                        cmd = command;
+                    }
+                }
```

Comment:
Why do we do this though? isn't this what `commandsPerCommand` is for?

Reply:
Ah, I should remove `commandsPerCommand`. Turns out that we cannot do anything so nice because we have distinct commands with the same name (but different help / behavior), so a command is not a single word but rather a prefix.